llpc/context/llpcCompiler.cpp

```diff
@@ -1418,12 +1429,162 @@ Result Compiler::buildGraphicsPipelineInternal(GraphicsContext *graphicsContext,
+// Build graphics pipeline using part-pipeline compilation
+//
+// @param graphicsContext : Graphics context this graphics pipeline
+// @param shaderInfo : Shader info of this graphics pipeline
+// @param [out] pipelineElf : Output Elf package
+// @param [out] stageCacheAccesses : Stage cache access result. All elements must be initialized in the caller as
+//                                   CacheAccessInfo::CacheNotChecked
+Result Compiler::buildGraphicsPipelineWithPartPipelines(Context *context,
+                                                        ArrayRef<const PipelineShaderInfo *> shaderInfo,
+                                                        ElfPackage *pipelineElf,
+                                                        MutableArrayRef<CacheAccessInfo> stageCacheAccesses) {
+  Result result = Result::Success;
+  unsigned wholeStageMask = context->getShaderStageMask();
+  ElfPackage partPipelineBuffers[2];
+  std::unique_ptr<Pipeline> elfLinkerPipeline;
+  std::unique_ptr<ElfLinker> elfLinker;
+
+  // For each of the two part pipelines (fragment first, then pre-rasterization)...
+  for (unsigned partIdx = 0; partIdx != 2; ++partIdx) {
+    // Compile the part pipelines, or find it in the cache.
+    llvm::StringRef partPipelineElf;
+    unsigned partStageMask = partIdx == 0 ? 1 << ShaderStageFragment : wholeStageMask & ~(1 << ShaderStageFragment);
+    context->getPipelineContext()->setShaderStageMask(partStageMask);
+
+    if (partIdx == 0 && (wholeStageMask & (1 << ShaderStageGeometry)))
+      context->getPipelineContext()->setPreRastHasGs(true);
+
+    // Hash the selected shaders and the pipeline state applicable to them.
+    Util::MetroHash64 hasher;
+    for (unsigned shaderIndex = 0; shaderIndex < shaderInfo.size() && result == Result::Success; ++shaderIndex) {
+      if (const PipelineShaderInfo *shaderInfoEntry = shaderInfo[shaderIndex]) {
+        ShaderStage stage = shaderInfoEntry->entryStage;
+        if (shaderInfoEntry->pModuleData && (partStageMask & (1 << stage))) {
+          // This is a shader to include in this part pipeline. Add the shader code and options to the hash.
+          PipelineDumper::updateHashForPipelineShaderInfo(stage, shaderInfoEntry, /*isCacheHash=*/true, &hasher,
+                                                          /*isRelocatableShader=*/false);
+        }
+      }
+    }
+    // Add applicable pipeline state to the hash. (This uses getShaderStageMask() to decide which parts of the
+    // state are applicable.)
+    context->getPipelineContext()->setPipelineState(/*pipeline=*/nullptr, /*hasher=*/&hasher, /*unlinked=*/false);
+
+    // If we already have an ELF linker pipeline object, we must be doing the pre-rasterization part pipeline,
+    // and the linker's pipeline object contains the FS input mapping state. We need to include that state in
+    // the hash.
+    if (elfLinker) {
```

Comment:
If all this is really checking is if the pre-rasterization part is being compiled, then actaully check that by checking partIdx.  Then a comment would not be needed.

Reply:
Changed to check the stage instead.